swf/src/types/fixed.rs

```diff
@@ -0,0 +1,643 @@
+            #[inline]
+            pub fn wrapping_div(self, other: Self) -> Self {
+                let n = (<$intermediate_type>::from(self.0) << Self::FRACTIONAL_BITS).wrapping_div(<$intermediate_type>::from(other.0));
+                Self(n as $underlying_type)
+            }
+
+            /// Wrapping (modular) multiplication.
+            /// Multiplies this fixed-point by an integer, returning the integer result.
+            /// The result will use the full size of the integer. The fractional bits will be truncated.
+            #[inline]
+            pub fn wrapping_mul_int(self, other: $underlying_type) -> $underlying_type {
+                let n = (<$intermediate_type>::from(self.0)
+                    .wrapping_mul(<$intermediate_type>::from(other)))
+                    >> Self::FRACTIONAL_BITS;
+                n as $underlying_type
+            }
+        }
+
+        impl Default for $type_name {
+            /// Returns the default value of `0.0`.
+            #[inline]
+            fn default() -> Self {
+                Self(0)
+            }
+        }
+
+        impl std::fmt::Display for $type_name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}", self.to_f64())
+            }
+        }
+
+        impl Neg for $type_name {
+            type Output = Self;
+            #[inline]
+            fn neg(self) -> Self {
+                Self(-self.0)
+            }
+        }
+
+        // fixed + fixed
+        impl Add for $type_name {
+            type Output = Self;
+
+            #[inline]
+            fn add(self, other: Self) -> Self {
+                Self(self.0 + other.0)
+            }
+        }
+
+        // fixed += fixed
+        impl AddAssign for $type_name {
+            #[inline]
+            fn add_assign(&mut self, other: Self) {
+                *self = *self + other
+            }
+        }
+
+        // fixed - fixed
+        impl Sub for $type_name {
+            type Output = Self;
+
+            #[inline]
+            fn sub(self, other: Self) -> Self {
+                Self(self.0 - other.0)
+            }
+        }
+
+        // fixed -= fixed
+        impl SubAssign for $type_name {
+            #[inline]
+            fn sub_assign(&mut self, other: Self) {
+                *self = *self - other
+            }
+        }
+
+        // fixed * fixed
+        impl Mul for $type_name {
+            type Output = Self;
+
+            #[inline]
+            fn mul(self, other: Self) -> Self {
+                let n = <$intermediate_type>::from(self.0) * <$intermediate_type>::from(other.0)
+                    >> Self::FRACTIONAL_BITS;
+                if cfg!(debug_assertions) {
+                    Self(<$underlying_type>::try_from(n).expect("Attempted to multiply with overflow"))
+                } else {
+                    Self(n as $underlying_type)
+                }
+            }
```

Comment:
Should this delegate to `mul_int`?
It's not that much code, so not big deal, but the body looks completely identical to that.
Same for `div` and `div_int`.

Maybe it would be just confusing, because the same amount of right shift (for `mul` that is) is needed for a different reason?
I mean, in `mul_int`, it converts the result to "regular integer"; while in here, it "undoes" one of the two incoming "shifted representations". I hope it's clear what I'm trying to say.

Reply:
Yea, that makes sense. It's slightly awkward as I'd have to to do `self.mul(Fixed8::from_bits(other)).0` which isn't super obvious. More generally we could probably have a single method for each op that both the normal and wrapping methods defer to. I'll leave it as is for now.